File: Lab/SEM4/MPP/lab9/moto_java/common/src/main/java/org/zdroba/repository/UserRepositoryImpl.java
package org.zdroba.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.zdroba.entity.User;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class UserRepositoryImpl implements UserRepository {
    private final Connection connection  = DataBaseConnection.getInstance().getConnection();
    private static final Logger logger = LogManager.getLogger();

    @Override
    public User find(Long id) {
        logger.traceEntry();
        String sql = "SELECT * FROM users WHERE id = ? LIMIT 1";

        try(PreparedStatement ps = connection.prepareStatement(sql)){
            ps.setLong(1, id);

            try(ResultSet rs = ps.executeQuery()){

                if(rs.next()){
                    String email = rs.getString("email");
                    String password = rs.getString("password");

                    User user = new User(email, password);
                    user.setId(id);

                    logger.traceExit();
                    return user;
                }
            }
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
        return null;
    }

    @Override
    public User find(String email) {
        logger.traceEntry();
        String sql = "SELECT * FROM users WHERE email = ? LIMIT 1";

        try(PreparedStatement ps = connection.prepareStatement(sql)){
            ps.setString(1, email);

            try(ResultSet rs = ps.executeQuery()){

                if(rs.next()){
                    Long id = rs.getLong("id");
                    String password = rs.getString("password");

                    User user = new User(email, password);
                    user.setId(id);

                    logger.traceExit();
                    return user;
                }
            }
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
        return null;
    }

    @Override
    public List<User> getAll() {
        logger.traceEntry();
        String sql = "SELECT * FROM users";

        List<User> users = new ArrayList<>();

        try(PreparedStatement ps = connection.prepareStatement(sql)){
            try(ResultSet rs = ps.executeQuery()){

                while(rs.next()){
                    Long id = rs.getLong("id");
                    String email = rs.getString("email");
                    String password = rs.getString("password");

                    User user = new User(email, password);
                    user.setId(id);

                    users.add(user);
                }
            }
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
        return users;
    }

    @Override
    public void add(User user) {
        logger.traceEntry("Saving User{}", user);
        String sql = "INSERT INTO users (email, password) VALUES (?,?)";

        try(PreparedStatement ps = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)){

            ps.setString(1, user.getEmail());
            ps.setString(2, user.getPassword());

            int result = ps.executeUpdate();

            ResultSet rs = ps.getGeneratedKeys();
            if (rs.next()) {
                user.setId(rs.getLong(1));
            }

            logger.trace("Saved {} instance", result);
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
    }
}

File: Lab/SEM4/MPP/lab11/moto_c/src/Common/repository/UserRepositoryImpl.cs
using moto_c.entity;
using MySqlConnector;
using NLog;

namespace moto_c.repository;

public class UserRepositoryImpl: UserRepository
{

    private readonly MySqlConnection _connection = DatabaseConnection.getInstance().GetConnection();
    private static readonly Logger logger = LogManager.GetCurrentClassLogger();
    
    public User? find(long id)
    {
        logger.Trace("Entry Find with id={}", id);
        string sql = "SELECT * FROM users WHERE id = ? LIMIT 1";

        try
        {
            using (MySqlCommand cmd = new MySqlCommand(sql, _connection))
            {
                cmd.Parameters.AddWithValue("@id", id);

                using (MySqlDataReader rs = cmd.ExecuteReader())
                {
                    if (rs.Read())
                    {
                        string email = rs.GetString("email");
                        string password = rs.GetString("password");

                        User user = new User(email, password);
                        user.id = id;
                        
                        logger.Trace("Exit Find with id={}", id);
                        return user;
                    }
                }
            }
        }
        catch (MySqlException e)
        {
            logger.Error(e.Message);
            Console.Error.WriteLine(e.Message);
        }

        logger.Trace("Exit Find with id={}", id);
        return null;
    }

    public User? find(string email)
    {
        logger.Trace("Entry Find with email={}", email);
        string sql = "SELECT * FROM users WHERE email = ? LIMIT 1";

        try
        {
            using (MySqlCommand cmd = new MySqlCommand(sql, _connection))
            {
                cmd.Parameters.AddWithValue("@email", email);

                using (MySqlDataReader rs = cmd.ExecuteReader())
                {
                    if (rs.Read())
                    {
                        long id = rs.GetInt64("id");
                        string password = rs.GetString("password");

                        User user = new User(email, password);
                        user.id = id;
                        
                        logger.Trace("Exit Find with email={}", email);
                        return user;
                    }
                }
            }
        }
        catch (MySqlException e)
        {
            logger.Error(e.Message);
            Console.Error.WriteLine(e.Message);
        }

        logger.Trace("Exit Find with email={}", email);
        return null;
    }

    public List<User> getAll()
    {
        logger.Trace("Entry GetAll");
        string sql = "SELECT * FROM users";
        List<User> users = new List<User>();
        
        try
        {
            using (MySqlCommand cmd = new MySqlCommand(sql, _connection))
            {
                using (MySqlDataReader rs = cmd.ExecuteReader())
                {
                    while (rs.Read())
                    {
                        long id = rs.GetInt64("id");
                        string email = rs.GetString("email");
                        string password = rs.GetString("password");

                        User user = new User(email, password);
                        user.id = id;
                        users.Add(user);
                    }
                }
            }
        }
        catch (MySqlException e)
        {
            logger.Error(e.Message);
            Console.Error.WriteLine(e.Message);
        }
        
        logger.Trace("Exit GetAll");
        return users;
    }

    public void add(User user)
    {
        logger.Trace("Entry Add with id={}", user.id);
        string sql = "INSERT INTO users (email, password) VALUES (@email, @password)";

        try
        {
            using (MySqlCommand cmd = new MySqlCommand(sql, _connection))
            {
                cmd.Parameters.AddWithValue("@email", user.email);
                cmd.Parameters.AddWithValue("@password", user.password);

                int result = cmd.ExecuteNonQuery();

                user.id = cmd.LastInsertedId;
                logger.Trace("Saved {} instance", result);
            }
        }
        catch (MySqlException e)
        {
            logger.Error(e.Message);
            Console.Error.WriteLine(e.Message);
        }
        
        logger.Trace("Exit Add with id={}", user.id);
    }
}